cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -986,6 +989,58 @@ public ApplicationWithPrograms deployApp(NamespaceId namespace, @Nullable String
       Map<String, String> userProps)
       throws Exception {
     // TODO CDAP-19828 - remove appVersion parameter from method signature
+    return deployApp(namespace, appName, appVersion, summary, configStr, changeSummary,
+        sourceControlMeta, programTerminator, ownerPrincipal, updateSchedules, isPreview,
+        false, userProps);
+  }
+
+  /**
+   * Deploy an application using the specified artifact and configuration. When an app is deployed,
+   * the Application class is instantiated and configure() is called in order to generate an {@link
+   * ApplicationSpecification}. Programs, datasets, and streams are created based on the
+   * specification before the spec is persisted in the {@link Store}. This method can create a new
+   * application as well as update an existing one (when LifecycleManagement feature is disabled).
+   *
+   * @param namespace the namespace to deploy the app to
+   * @param appName the name of the app. If null, the name will be set based on the application
+   *     spec
+   * @param summary the artifact summary of the app
+   * @param configStr the configuration to send to the application when generating the
+   *     application specification
+   * @param changeSummary the change summary entered by the user - includes the description and
+   *     parent-version
+   * @param sourceControlMeta the source control metadata of an application that is pulled from
+   *     linked git repository
+   * @param programTerminator a program terminator that will stop programs that are removed when
+   *     updating an app. For example, if an update removes a flow, the terminator defines how to
+   *     stop that flow.
+   * @param ownerPrincipal the kerberos principal of the application owner
+   * @param updateSchedules specifies if schedules of the workflow have to be updated, if null
+   *     value specified by the property "app.deploy.update.schedules" will be used.
+   * @param isPreview whether the app deployment is for preview
+   * @param skipMakingLatest if true, the deployed app is not marked as latest
+   * @param userProps the user properties for the app deployment, this is basically used for
+   *     preview deployment
+   * @return information about the deployed application
+   * @throws InvalidArtifactException if the artifact does not contain any application classes
+   * @throws IOException if there was an IO error reading artifact detail from the meta store
+   * @throws ArtifactNotFoundException if the specified artifact does not exist
+   * @throws Exception if there was an exception during the deployment pipeline. This exception
+   *     will often wrap the actual exception
+   */
+  public ApplicationWithPrograms deployApp(NamespaceId namespace, @Nullable String appName,
```

Comment:
this is too many deployApp() methods, should consolidate them by putting the optional parameters in some DeployOptions class or something like that.

Reply:
At the very least it should be an additional argument to an existing method. We should not be adding two new deploy methods.

Reply:
Removed the overloaded methods that I had added previously. Instead added the new param to existing deploy methods and updated the places where they are used.